File: mobile-app/lib/core/utils/course_rating.dart
/// Deterministic pseudo-random ratings per course (stable across sessions).
class CourseRating {
  CourseRating._();

  static const double minRating = 4.0;
  static const double maxRating = 5.0;

  /// Returns a fixed rating between 4.0 and 5.0 (0.1 steps) for [courseId].
  static double forCourse(String courseId) {
    if (courseId.isEmpty) return 4.5;
    final h = _hash(courseId);
    final step = h % 11; // 0..10 → 4.0..5.0
    return minRating + step / 10;
  }

  /// Fixed review count for social proof (stable per course).
  static int reviewCountFor(String courseId) {
    if (courseId.isEmpty) return 120;
    final h = _hash('${courseId}_reviews');
    return 95 + (h % 2891); // 95 – 2985
  }

  static String formatCount(int count) {
    if (count >= 1000) {
      final k = count / 1000;
      return k >= 10 ? '${k.round()}k' : '${k.toStringAsFixed(1)}k';
    }
    return '$count';
  }

  static int _hash(String value) {
    var hash = 0;
    for (final unit in value.codeUnits) {
      hash = (hash * 31 + unit) & 0x7fffffff;
    }
    return hash;
  }
}

File: frontend/src/utils/courseRating.js
export function hashString(value) {
  let hash = 0
  for (let i = 0; i < value.length; i += 1) {
    hash = (hash * 31 + value.charCodeAt(i)) & 0x7fffffff
  }
  return hash
}

export function courseRating(courseId) {
  if (!courseId) return 4.5
  const step = hashString(courseId) % 11
  return 4.0 + step / 10
}

export function courseReviewCount(courseId) {
  if (!courseId) return 120
  return 95 + (hashString(`${courseId}_reviews`) % 2891)
}

export function formatReviewCount(count) {
  if (count >= 1000) {
    const k = count / 1000
    return k >= 10 ? `${Math.round(k)}k` : `${k.toFixed(1)}k`
  }
  return String(count)
}
